In [15]:
!pip install numpy
!pip install scipy
!pip install scipy.optimize

You should consider upgrading via the 'c:\users\user\appdata\local\programs\python\python37\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'c:\users\user\appdata\local\programs\python\python37\python.exe -m pip install --upgrade pip' command.


ERROR: Could not find a version that satisfies the requirement scipy.optimize (from versions: none)
ERROR: No matching distribution found for scipy.optimize
You should consider upgrading via the 'c:\users\user\appdata\local\programs\python\python37\python.exe -m pip install --upgrade pip' command.


In [16]:
import numpy as np
import scipy as sp
from scipy.optimize import brentq

<br>Величина кредита - <b>pv</b><br>Баланс после всех платежей - <b>fv = 0</b>

In [17]:
pv = 100000
fv = 0

<br>Величина периодического платежа (1 вариант) - <b>pmt1</b><br>Количество платежей (1 вариант) - <b>n1</b>

In [18]:
pmt1 = -1500
n1 = 120

<br>Величина периодического платежа (2 вариант) - <b>pmt2</b><br>Количество платежей (2 вариант) - <b>n2</b>

In [19]:
pmt2 = -1250
n2 = 180

<br>Величина периодического платежа (3 вариант) - <b>pmt3</b><br>Количество платежей (3 вариант) - <b>n3</b>

In [20]:
pmt3 = -1160
n3 = 240

In [21]:
def fun1(i):
    return pv*(1+i)**n1 + pmt1*(((1+i)**n1-1)/i) + fv 
def fun2(i):
    return pv*(1+i)**n2 + pmt2*(((1+i)**n2-1)/i) + fv 
def fun3(i):
    return pv*(1+i)**n3 + pmt3*(((1+i)**n3-1)/i) + fv 

In [22]:
sol1 = brentq(fun1, 0.01, 1.0)
sol1

0.010930603939868333

In [23]:
sol2 = brentq(fun2, 0.01, 1.0)
sol2

0.01063980717924384

In [24]:
sol3 = brentq(fun3, 0.01, 1.0)
sol3

0.010697698792848072

<br>Функция, решающая уравнение $ PV*(1+i)^n + PMT*\frac{(1+i)^n-1}{i} + FV = 0 $:

In [25]:
def solve_fun(pv, pmt, n, fv):
    def fun(i):
        return pv*(1+i)**n + pmt*((((1+i)**n)-1)/i) + fv
    return brentq(fun, 0.001, 1.0)

<br>Процентная ставка (1 вариант) - <b>i1</b>

In [26]:
i1 = solve_fun(pv, pmt1, n1, fv)
print(i1)

0.010930603939676818


<br>Процентная ставка (2 вариант) - <b>i2</b>

In [27]:
i2= solve_fun(pv, pmt2, n2, fv)
print(i2)

0.01063980717901046


<br>Процентная ставка (3 вариант) - <b>i3</b>

In [28]:
i3= solve_fun(pv, pmt3, n3, fv)
print(i3)

0.010697698792974464


<br>Годовые процентные ставки - <b>iy</b><sub>n</sub>

In [29]:
iy1 = 12*i1
iy2 = 12*i2
iy3 = 12*i3
print("I1year=", iy1*100,'%')
print("I2year=", iy2*100,'%')
print("I3year=", iy3*100,'%')

I1year= 13.11672472761218 %
I2year= 12.767768614812553 %
I3year= 12.837238551569358 %


<h1>Вывод:</h1>
<div>
    Для решения был использован метод Брента как наиболее удобный, универсальный и популярный.
    <ul>
        <li>13.12% для варианта <b>1</b></li>
        <li>12.77% для варианта <b>2</b></li>
        <li>12.84% для варианта <b>3</b></li>
    </ul>    
</div>
Наиболее хорошая годовая ставка оказалась под номером 2.